test/binary.carp

```diff
@@ -0,0 +1,145 @@
+(use Binary)
+
+(load "Test.carp")
+(use Test)
+
+(deftest test
+  ;; int16 tests
+  (assert-equal test
+                (Uint16.from-long 2051l)
+                (unsafe-bytes->int16 (Binary.Order.LittleEndian) &[3b 8b])
+                "Unsafe little endian unsafe-bytes->int16 works as expected.")
+
+  (assert-equal test
+                (Uint16.from-long 776l)
+                (unsafe-bytes->int16 (Binary.Order.BigEndian) &[3b 8b])
+                "Unsafe big endian unsafe-bytes->int16 works as expected.")
+
+  (assert-equal test
+                &(Maybe.Just (Uint16.from-long 2051l))
+                &(bytes->int16 (Binary.Order.LittleEndian) &[3b 8b])
+                "Unsafe little endian bytes->int16 works as expected.")
+
+  (assert-equal test
+                &(Maybe.Just (Uint16.from-long 776l))
+                &(bytes->int16 (Binary.Order.BigEndian) &[3b 8b])
+                "Unsafe big endian bytes->int16 works as expected.")
+
+  (assert-equal test
+                &(Maybe.Nothing)
+                &(bytes->int16 (Binary.Order.LittleEndian) &[3b])
+                "bytes->int16 returns Nothing on insufficient data.")
+
+  (assert-equal test
+                &[(Uint16.from-long 2051l) (Uint16.from-long 776l)]
+                &(unsafe-bytes->int16-seq (Binary.Order.LittleEndian) &[3b 8b 8b 3b])
+                "Unsafe little endian bytes->int16-seq works as expected.")
+
+  (assert-equal test
+                &[(Maybe.Just (Uint16.from-long 776l)) (Maybe.Just (Uint16.from-long 2051l)) (Maybe.Nothing)]
```

Comment:
I’m not entirely sure this API is useful as is (since in order to check whether everything was decoded, you’ll have to check whether the last element is `Maybe.Nothing`). Maybe making the entire thing a `Maybe` makes more sense?

Reply:
A pretty typical idiom is to return the read value(s) + number of bytes read--or, alternatively, the number of excess bytes not read-- perhaps that works? Then we'd have:

`bytes->intN-seq (Fn [Order (Array Byte)] (Pair (Array UintN) Int))`

I think your suggestion works too. I'm open to any of the following signatures:

`(Fn [Order (Array Byte)] (Pair (Array UintN) Int))` -- some + number of bytes read or not read.
`(Fn [Order (Array Byte)] (Maybe (Array UintN))` -- all or nothing.
`(Fn [Order (Array Byte)] (Result (Array Byte))` -- all or nothing just like the maybe variant, but return the number of excess bytes for the failure case '

@hellerve @eriksvedang any preference? Of course, we could just support all variants--I could see each being useful under different circumstances.

Reply:
The number of bytes read is somewhat redundant, since we have a length property on the array, I think. I think I personally like the last version most; it seems like a local optimum for both safety and information content.

Reply:
> The number of bytes read is somewhat redundant, since we have a length property on the array, I think. I think I personally like the last version most; it seems like a local optimum for both safety and information content.

Fair point! 

I think returning the number read is a convenience thing, since the length of the resulting int array != the length of the number of read bytes. e.g. reading bytes into an array of 2 int16s requires an input of 4 bytes, reading into 2 int32s requires 8 bytes--the calculation is trivial, but it is convenient to just get it back automatically. It's more useful for cases where you have some N bytes and just want to read say, as many int16s as you can--if you have 5 bytes, you can only read 2 int16s, which is easy enough to figure out...but it's more convenient to just get the number from the function.

Still, I think bytes remaining (not read) is more useful information anyway for things like scanning files or working with streams--after some thought, I think both the `Pair` and `Result` variants are useful for different use cases. When you're scanning or streaming it's kind of painful to have to get your numbers exactly correct, and you often want to know how many bytes remain in order to determine how to/if you can continue processing the data (use `Pair`). In other circumstances, being unable to process some precise sequence of bytes is in fact catastrophic, and we can't really continue with our task (use `Result`).

For example, I'm going to use this to read terminfo files--the first section of those files are called a header, which is made up of 6 littleEndian int16s (12 bytes). 

Using `(Fn [Order (Array Byte)] (Pair (Array UintN) Int))` where the Int returned is bytes remaining, it'd be possible to read the entire terminfo file into a single byte array, read the first 12 bytes, and use `Pair.b` to see how many bytes remain in the file, and process those accordingly. This is convenient for a "greedy" approach in which we eagerly read in a file.

Contrarily, using `(Fn [Order (Array Byte)] (Result (Array Byte) Int)` still works, but instead of working with the entirety of the file, we need to be specific about the bytes we need and process each piece separately--if we read the whole file into a single byte array and passed it to this function, it'd result in a failure if it was longer than just the 12byte header. This is convenient for a stream based or "incremental" approach to processing the file.

I feel like the latter `Result` based function is appropriate when failed reads are not recoverable--for instance, in my example, if terminfo files *must* have a header, it makes sense to use the `Result` variant and stop if we fail.

On the other hand, the `Pair` variant is good for cases in which a failure is not catastrophic and you want to just keep chugging along processing all the data you can, or do something with the data remaining.

So, I'm a fan of both. I think the `Pair` version could be defined in terms of the `Result` version--so that may be the only one we need.

Sorry for the long post, wdyt?

Reply:
I think having both makes a lot of sense. I follow your reasoning completely: sometimes failed reads are recoverable, sometimes they aren’t, and having a specialized function for either work makes sense.

I’m not sure how the `Pair` version could be defined in terms of the other version; shouldn’t it be the other way around?

Reply:
I believe either can be taken as the primary definition because sum and product types are dual, and these types model sum and product types (pair being product and result being sum) (in categorical mumbo jumbo, pair is the product of some category C while Result is the coproduct).

...as to what actually makes the most sense in practice, I have no idea--I'll try to find the clearest solution. I'll update this PR later today!

Thanks for your comments! 

Thinking further about this, it would also be helpful to have a function that just returned the rest of the byte array (the unprocessed bytes `Fn [Order (Array Byte)] (Pair (Array Uint) (Array Byte))`--this is the shape of a lot of stream based scanning/parsing techniques, where parsing the next N items in the stream returns the parsed value + the remainder of the stream). Functions that let you decide how many bytes you want to process in a given byte array should also be useful--but I'll add these in later PRs.

Reply:
You were right, @hellerve! Defining `Results` in terms of `Pairs` is simpler

Reply:
Yay! Thanks for trying it out!